lib/src/Uncertainty/Bayesian/ThreeDVAR.cxx

```diff
@@ -0,0 +1,439 @@
+	{
+	  const Matrix parameterGradient(parametrizedModel.parameterGradient(inputObservations[i]));
+	  const Matrix scaledParameterGradient(parameterGradient * errorInverseCholesky);
+	  std::copy(scaledParameterGradient.getImplementation()->begin(), scaledParameterGradient.getImplementation()->end(), gradientObservations.begin() + shift);
+	  shift += skip;
+	}
+      for (UnsignedInteger j = 0; j < parameterDimension; ++j)
+	for (UnsignedInteger i = 0; i <= j; ++i)
+	  gradientObservations(i, size * outputDimension + j) = parameterInverseCholesky(i, j);
+      return gradientObservations;
+    }
+
+    UnsignedInteger getInputDimension() const
+    {
+      return evaluation_.getInputDimension();
+    }
+
+    UnsignedInteger getOutputDimension() const
+    {
+      return evaluation_.getOutputDimension();
+    }
+
+    Description getInputDescription() const
+    {
+      return evaluation_.getInputDescription();
+    }
+
+    Description getOutputDescription() const
+    {
+      return evaluation_.getOutputDescription();
+    }
+
+    Description getDescription() const
+    {
+      return evaluation_.getDescription();
+    }
+
+    String __repr__() const
+    {
+      OSS oss;
+      oss << "CalibrationModelGradient(" << evaluation_ << ")";
+      return oss;
+    }
+
+    String __str__(const String & ) const
+    {
+      OSS oss;
+      oss << "CalibrationModelGradient(" << evaluation_ << ")";
+      return oss;
+    }
+
+  private:
+    const CalibrationModelEvaluation evaluation_;
+
+  }; // class CalibrationModelGradient
+
+}
+
+/* Performs the actual computation. Must be overloaded by the actual calibration algorithm */
+void ThreeDVAR::run()
+{
+  // Compute the posterior MAP
+  // Error distribution
+  const Normal error(Point(errorCovariance_.getDimension()), errorCovariance_);
+  const TriangularMatrix parameterInverseCholesky(getParameterPrior().getInverseCholesky());
+  const TriangularMatrix errorInverseCholesky(error.getInverseCholesky());
+  const Point thetaStar(run(inputObservations_, outputObservations_, getCandidate(), parameterInverseCholesky, errorInverseCholesky));
+  // Compute the posterior distribution
+  Distribution parameterPosterior;
+  if (bootstrapSize_ > 0)
+    {
+      // Compute the covariance using Bootstrap
+      Sample joinedData(inputObservations_);
+      joinedData.stack(outputObservations_);
+      const BootstrapExperiment bootstrap(joinedData);
+      Sample thetaSample(bootstrapSize_, thetaStar.getDimension());
+      Indices inputIndices(inputObservations_.getDimension());
+      inputIndices.fill();
+      Indices outputIndices(outputObservations_.getDimension());
+      outputIndices.fill(inputIndices.getSize());
+      Sample empty;
+      for (UnsignedInteger i = 0; i < bootstrapSize_; ++i)
+	{
+	  const Sample joinedSample(bootstrap.generate());
+	  thetaSample[i] = run(inputObservations_, outputObservations_, thetaStar, parameterInverseCholesky, errorInverseCholesky);
```

Comment:
run(joinedSample.getMarginal(inputIndices), joinedSample.getMarginal(outputIndices), thetaStar, parameterInverseCholesky, errorInverseCholesky);

Reply:
Well done! It must be the root cause of the test failure!